keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,908 @@
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No.
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No.
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+- Users using the new field will see an see an increase in the Service object of
+  ~10 bytes
+- Also, in situations when the status for this new field is reported, size
+  equivalent to the addition of two Condition types may be observed.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No.
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No.
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+No.
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+_This section will be completed when targeting beta to a release._
+
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
```

Comment:
Not sure that folks are going to like it, but: I'd like to see an alternative discussed where the appropriate heuristic to use is captured in a new API kind.

Let's call this TargetPodHeuristic for now.

So a Service might look like:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: example-service
  namespace: default
spec:
  heuristics:
    - example-heuristic
  selector:
    app.kubernetes.io/name: example
  ports:
    - protocol: TCP
      port: 24
      targetPort: 24
```

and the TargetPodHeuristic for that:
```yaml
apiVersion: example/v1alpha1
kind: TargetPodHeuristic
metadata:
  name: example-heuristic
  namespace: default
spec:
    preferredTargets:
      - weight: 23
        matchLabelKeys:
         - topology.kubernetes.io/zone   # prefer a target in the same zone
         - topology.kubernetes.io/region # prefer a target in the same region
                                         # maybe zone names are only regionally unique
      - weight: 11
        matchLabelKeys:
         - topology.kubernetes.io/region # prefer a target in the same region
    avoidedTargets:
      - weight: 42
        matchTaints:
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoSchedule"
        - key: "node.kubernetes.io/unreachable"
          operator: "Exists"
          effect: "NoExecute"
```

To be clear: we can rule this out if we want to. That's absolutely fine. What I'm calling for is enough of a write-up about why it wouldn't fit / couldn't work / doesn't scale / etc. Let's not leave readers to guess this themselves.

Reply:
sftim -- That's a reasonable ask. Thanks. I've tried to add something. Hope that conveys the right ideas.

Reply:
This skews back to `topologyKeys` which gave users a ton of control but made alternative implementations very difficult because it was over-precise.

It's a fine line we are walking between defining a full service mesh API and defining too little to be useful.